File: ChoiceTheory/Matlab/CPT.m
% Cumulative Prospect Theory function (CPT v2.0)
%% Introduction
% v1.0 is to tackle 2 outcomes in a prospect only
% v2.0 indicates an advanced version of CPT v1.0
% It is a generalised model for n outcomes in a prospect (so far I have
% tested with 3 outcomes in a prospect from exercise in week 5)
% Includes CPT, v and w function
% Derive the idea from the loop function in PS923 Assignment 3 (RFT)
% Combine probabilities and outcomes of a prospect in terms of VECTOR
%% Function
function sv = CPT2(p,X,alpha,gamma,lambda,beta,delta)
%% Create empty vectors for storing values
vp=zeros(1,size(X,2)); pp=zeros(1,size(X,2));
vn=zeros(1,size(X,2)); pn=zeros(1,size(X,2));
vpf=zeros(1,size(X,2)); vnf=zeros(1,size(X,2));
%% Pre-test checking
% Check if the length of outcomes and probabilites within a prospect are
% the same
% numel(X) = number of array elements in X
    if numel(X) ~= numel(p)
        disp('Outcomes and Probabilities must have the same length!')
    end
    
% Check if the sum of probabilites is equal to 1   
    if sum(p) ~= 1
        disp('Please revise your probability vector!')
    end
    
% Check if the outcomes are in an ascending order (using issorted(X))
    if issorted(X) == 0
        [X,idx]=sort(X);
        p=p(:,idx);
    end
%% For loop
    for m=1
        limit=size(X,2);
        while (m <= limit)
            if X(m) >= 0 
                vp(m) = X(m).^(alpha); % Value obtained when X(m)>=0
                pp(m) = w(sum(p(m:limit)),gamma) - w(sum(p(m+1:limit)),gamma); % Decision weights obtained
                vpf(m) = sum(vp(m).*pp(m)); % subjective value of a positive outcome
            else
                vn(m) = -lambda*(abs(X(m))).^(beta); % Value obtained when X(m)<0
                pn(m) = w(sum(p(1:m)),delta) - w(sum(p(1:m-1)),delta); % Decision weights obtained
                vnf(m)= sum(vn(m).*pn(m)); % subjective value of a negative outcome
            end
            m = m + 1;
            sv = sum(vpf)+sum(vnf); % subjective value after combining values of both positive and negative outcomes
        end
    end
end
%% Weighting function
% two inputs
% p = probabilities of outcomes within a prospect (vector)
% c = free parameter assigning weights according to p
% c = gamma when the outcome is positive
% c = delta when the outcome is negative
% weight = weight of a probability under weighting function
function weight = w(p,c)
weight= (p.^c) / ((p.^c+(1-p).^c)^(1/c));
end
%% Value function
% three inputs
% x = outcome within a prospect (vector)
% exp = parameter assigning a value according to x
% exp = alpha if x >= 0
% exp = beta if x < 0 
% lambda = parameter showing losses loom larger than gain (loss aversion)
function value=v(X,exp,lambda)
if X>=0
    value = X.^exp;
else
    value = -lambda(abs(X)).^exp;
end
end

File: ChoiceTheory/R/CPT.R
# Materials (Options A and B)
# outcomes must be shown in decreasing order (vector)
x_A <-c(200, 22, 20) 
p_A <-c(0.9, 0.05, 0.05)
x_B <-c(200, 190, 20)
p_B <-c(0.85, 0.05, 0.1)

#Cumulative Prospect Theory (CPT)
alpha <- 0.88
gamma <- 0.65
lamda <- 2.25
beta <- 0.88
delta <- 0.69

# CPT function
# Weighting and value function
w <- function (p,c) {p^c / ((p^c+(1-p)^c)^(1/c))}
v <- function(x,a,lambda) ifelse(x >= 0,x^a,-lambda*abs(x)^a)

CPT <- function (p,X,alpha,gamma,lambda,beta,delta){
# Pre-test checking
if (length(X) != length(p)){ # Check if X and p have the same number of elements
	print('Outcomes and probabilities should have the same length!')
}
if (is.unsorted(X)) { # Check if the outcomes are sorted in an increasing order
  oo <- order(X)
  X <- X[oo]
  p <- p[oo]
	print('Outcomes are in an increasing order. Proceed to CPT calculation')
}
if (sum(p) != 1){ # Check if the sum of probabilities is added up to 1
	print('Sum of probabilities must be added up to 1!')
}
# Create empty vectors for storing outputs
pp <- {} # pp = positive pi
np <- {} # np = negative pi
vp <- {} # vp = value of an outcome in a positive prospect
vn <- {} # vn = value of an outcome in a negative prospect
vpf <- {} # vpf = value of a positive prospect
vnf <- {} # vnf = value of a negative prospect
sv <- {} # sv = subjective value of a prospect

	for (i in 1: length(X)){
			if (X[i]>=0){
					vp[i] <- v(X[i],alpha,lambda)
					if (i < length(X)){
						pp[i] <- w(sum(p[i:length(p)]),gamma) - w(sum(p[(i+1):length(p)]),gamma)
					}else{
					pp[length(p)] <- w(p[length(p)],gamma)
					}
					vpf[i] <- sum(vp[i]*pp[i])	
			}else {
					vn[i] <- v(X[i],beta,lambda)
					np[i] <- w(sum(p[1:i]),delta) - w(sum(p[0:(i-1)]),delta)
					vnf[i] <- sum(vn[i]*np[i])	
			}		
	}
	vpf <- sum(vpf,na.rm = TRUE)
	vnf <- sum(vnf,na.rm = TRUE)
	sv <- vpf + vnf
	return (sv)
}


# Calculation
CPT_A <- CPT(p_A,x_A,alpha,gamma,lambda,beta,delta)
CPT_B <- CPT(p_B,x_B,alpha,gamma,lambda,beta,delta)

# Decision
if (CPT_A > CPT_B){
	print('Option A is chosen')
}else{
	print('Option B is chosen')
	}
